pyexchange/coinone.py

```diff
@@ -0,0 +1,256 @@
+        assert (isinstance(offset, int))
+
+        currency = pair.split('-')[0]
+
+        result = self._http_authenticated_request("POST", f"/v2/order/complete_orders", {"currency": currency})
+        return list(map(lambda item: CoinoneTrade.from_message(item, pair), result['completeOrders']))
+
+    def _choose_nonce(self) -> int:
+        with self.last_nonce_lock:
+            time.sleep(0.1)
+            timed_nonce = int(time.time() * 1000)
+
+            if self.last_nonce + 1 > timed_nonce:
+                self.logger.info(
+                    f"Wanted to use nonce '{timed_nonce}', but last nonce is '{self.last_nonce}', using '{self.last_nonce + 1}' instead")
+
+                self.last_nonce += 1
+            else:
+                self.last_nonce = timed_nonce
+
+            return self.last_nonce
+
+    def _get_encoded_payload(self, payload) -> bytes:
+        nonce = self._choose_nonce()
+
+        payload["access_token"] = self.access_token
+        payload["nonce"] = nonce
+
+        payload = json.dumps(payload, separators=(',', ':'))
+        return base64.b64encode(bytes(payload, 'utf-8'))
+
+    def _get_signature(self, encoded_payload) -> HMAC:
+        return hmac.new(bytes(self.secret_key, 'utf-8'), encoded_payload, hashlib.sha512).hexdigest()
+
+    def _http_authenticated_request(self, method: str, resource: str, body: dict):
+        assert (isinstance(method, str))
+        assert (isinstance(resource, str))
+        assert (isinstance(body, dict) or (body is None))
+
+        url = f"{self.api_server}{resource}"
+
+        encoded_payload = self._get_encoded_payload(body)
+
+        headers = {
+            "Accept": "Application/JSON",
+            'Content-type': 'application/json',
+            "X-COINONE-PAYLOAD": encoded_payload,
+            "X-COINONE-SIGNATURE": self._get_signature(encoded_payload)
+        }
+
+        return self._result(requests.request(method=method,
+                                             url=url,
+                                             data=encoded_payload,
+                                             headers=headers,
+                                             timeout=self.timeout))
+
+    def _http_unauthenticated_request(self, method: str, resource: str, body: dict):
+        assert (isinstance(method, str))
+        assert (isinstance(resource, str))
+        assert (isinstance(body, dict) or (body is None))
+
+        data = json.dumps(body, separators=(',', ':'))
+        url = f"{self.api_server}{resource}"
+
+        return self._result(requests.request(method=method,
+                                             url=url,
+                                             data=data,
+                                             timeout=self.timeout))
+
+    def _result(self, result) -> Optional[dict]:
+        if not result.ok:
+            raise Exception(f"Coinone API invalid HTTP response: {http_response_summary(result)}")
```

Comment:
Recommend not raising base exception types; `RuntimeError` may work better here and on line 253 below.

Reply:
Good point; updated accordingly.